### PR TITLE
use BTreeMap for empty objects with preserve-unknown-fields

### PR DESCRIPTION
**Pull request by alex-hunt-materialize**
If the object is empty and it has s-kubernetes-preserve-unknown-fields
set, do not create an empty struct.
Instead, use a `BTreeMap<String, serde_json::Value>`.

### CODE REVIEW

**Comment by clux** [comment]
I think the change in here looks great on paper, and most likely is going to work well. But I think we should have some tests here, so you could follow the basic flow:

- take your [linkerd crd](https://gist.github.com/alex-hunt-materialize/2743b1e2e58a49c4df0a11ecb39f46ab)
- strip annotations/labels/creation timestamps/uids/resource versions/generation
- write it to `tests/serverauth-crd.yaml`
- find an instance of that crd, strip secrets, labels etc, change `metadata.name` to `gen`, write to `tests/serverauth.yaml`
- extend `justfile` with a `test serverauth` target that applies both files, sets `CR = ServerAuthorization`, and add it to the ci workflow

at least that ensures the basic serialize/deserialize flow works.

However, I think it would be better to extract some relevant schema snippets from the crd (inlined as multiline comments passed to `serde_yaml`) so we can invoke the analyzer with that as unit tests for the analyzer (currently we have none of that sort), because that way we can more easily verify that more complicated properties hold (and it will be a bit easier to understand what is happening in this repo - the analyzer is all the complexity). WDYT?

**Comment by clux** [comment]
I have added an example of how I think we can do unit tests in https://github.com/kube-rs/kopium/commit/7085dfad93238b7c362ecb5b33e7a5f0aae6ab43 . They can be run locally with `cargo test --lib`, and the normal `RUST_LOG=debug` and `-- --nocapture` suffixes while debugging.

**Comment by alex-hunt-materialize** [comment]
That all sounds great. I'll try to get to it this week, but it might be Friday before I do.

**Comment by alex-hunt-materialize** [comment]
I've added a unit test and two justfile tests, but had to disable one. I have no idea why, but it fails with a 404 error when running the justfile test for the linkerd server. The resource exists, and it is able `get` it, but it can't find it when attempting to `patch` it.

File: tests/serverauth-crd.yaml [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    reason: NoConflicts
+    status: "True"
+    type: NamesAccepted
+  - lastTransitionTime: "2021-12-13T21:51:27Z"
+    message: the initial names have been accepted
+    reason: InitialNamesAccepted
+    status: "True"
+    type: Established
+  storedVersions:
+  - v1beta1
```
Comment (clux) [inline tests/serverauth-crd.yaml:311]: this status part should probably be removed, not part of the original crd

File: tests/server-crd.yaml [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    reason: NoConflicts
+    status: "True"
+    type: NamesAccepted
+  - lastTransitionTime: "2021-12-13T21:51:27Z"
+    message: the initial names have been accepted
+    reason: InitialNamesAccepted
+    status: "True"
+    type: Established
+  storedVersions:
+  - v1beta1
```
Comment (clux) [inline tests/server-crd.yaml:175]: similarly here, stray status object

File: justfile [hunk 3]
```diff
@@ -53,5 +53,21 @@ test-cluster:
...
+  kubectl apply -f tests/serverauth.yaml
+  cargo test --test runner -- --nocapture
+
+# This fails in the test runner when trying to patch the applied instance.
+# let _obj = cr.patch_status("gen", &pp, &patch).await?;
+#test-linkerd-server:
+#  kubectl apply --server-side -f tests/server-crd.yaml
+#  cargo run --bin kopium -- --docs servers.policy.linkerd.io > tests/gen.rs
+#  echo "pub type CR = Server;" >> tests/gen.rs
+#  kubectl apply -f tests/server.yaml
```
Comment (clux) [inline justfile:69]: you have run into an edge case of the test runner. if you name it something other than `server` the `path_status` branch won't happen.

..ideally, i should refactor that branch to only trigger if the file with the `gen` object has a status object, but never got that far.

**Review by clux — COMMENTED** [review]
looks great. tests are great, i did a merge resolution commit, and i think i messed up a brace.

have a suggestion for the failing integration test (which is mostly my bad on designing this ad-hoc test runner).

File: justfile (hunk 3 above)
Comment (alex-hunt-materialize) [reply justfile]: I just pushed a change to skip that if the file doesn't contain the status object.

File: tests/server-crd.yaml (hunk 2 above)
Comment (alex-hunt-materialize) [reply tests/server-crd.yaml]: done

File: tests/serverauth-crd.yaml (hunk 1 above)
Comment (alex-hunt-materialize) [reply tests/serverauth-crd.yaml]: done

**Review by clux — APPROVED** [review]

File: justfile (hunk 3 above)
Comment (clux) [reply justfile]: ah, that's great. thanks a lot!

**Comment by clux** [comment]
Pushing this and your previous PR into a 0.9.0 release. Thanks again for the help!

**Comment by alex-hunt-materialize** [comment]
No problem! Thanks for being so responsive!
